tools/testsys/src/aws_resources.rs

```diff
@@ -227,3 +297,181 @@ impl AwsK8s {
         Ok(Crd::Test(sonobuoy))
     }
 }
+
+/// In order to easily create migration tests for `aws-k8s` variants we need to implement
+/// `Migration` for it.
+impl Migration for AwsK8s {
+    fn migration_config(&self) -> Result<MigrationsConfig> {
+        Ok(MigrationsConfig {
+            tuf_repo: self
+                .tuf_repo
+                .as_ref()
+                .context("Tuf repo metadata is required for upgrade downgrade testing.")?
+                .clone(),
+            starting_version: self
+                .starting_version
+                .as_ref()
+                .context("You must provide a starting version for upgrade downgrade testing.")?
+                .clone(),
+            migrate_to_version: self
+                .migrate_to_version
+                .as_ref()
+                .context("You must provide a target version for upgrade downgrade testing.")?
+                .clone(),
+            region: self.region.to_string(),
+            secrets: self.secrets.clone(),
+            capabilities: self.capabilities.clone(),
+            assume_role: self.assume_role.clone(),
+        })
+    }
+
+    fn instance_provider(&self) -> String {
+        let cluster_name = self.cluster_name();
+        format!("{}-instances", cluster_name)
+    }
+
+    fn migration_labels(&self) -> BTreeMap<String, String> {
+        btreemap! {
+            "testsys/arch".to_string() => self.arch.to_string(),
+            "testsys/variant".to_string() => self.variant.to_string(),
+            "testsys/flavor".to_string() => "updown".to_string(),
+        }
+    }
+}
+
+/// An enum to differentiate between upgrade and downgrade tests.
+enum MigrationVersion {
+    ///`MigrationVersion::Starting` will create a migration to the starting version.
+    Starting,
+    ///`MigrationVersion::Migrated` will create a migration to the target version.
+    Migrated,
+}
+
+/// A configuration containing all information needed to create a migration test for a given
+/// variant.
+struct MigrationsConfig {
+    tuf_repo: TufRepoConfig,
+    starting_version: String,
+    migrate_to_version: String,
+    region: String,
+    secrets: Option<BTreeMap<String, SecretName>>,
+    capabilities: Option<Vec<String>>,
+    assume_role: Option<String>,
+}
+
+/// Migration is a trait that should be implemented for all traits that use upgrade/downgrade
+/// testing. It provides the infrastructure to easily create migration tests.
+trait Migration {
+    /// Create a migration config that is used to create migration tests.
+    fn migration_config(&self) -> Result<MigrationsConfig>;
+
+    /// Create the labels that should be used for the migration tests.
+    fn migration_labels(&self) -> BTreeMap<String, String>;
+
+    /// Return the name of the instance provider that the migration agents should use to get the
+    /// instance ids.
+    fn instance_provider(&self) -> String;
```

Comment:
This seems like a place where stronger typing could be useful. It sort of sticks out as having not very much to do with migration testing.

Reply:
I was wondering if an instance provider trait would be helpful or not.